calcom/cal.com
# feat: Validate and test OpenAPI spec

## What does this PR do?

This PR attempts to address testing the OpenAPI spec for consistency with the API by using Dredd.js. Dredd tests all listed responses including errors. Some of the low-hanging fruit to get some of the tests to pass were as simple as changing status codes or specifying the proper response body for a 401.

It does not solve all outstanding issues. I would be happy to work with someone from your team to help them understand how to fix the remaining issues. Some tests were skipped because testing them proves harder without modifying the seed script even more.

Current status:
`complete: 126 passing, 35 failing, 0 errors, 38 skipped, 199 total`

There is an issue with the production docs endpoint. I have commented about it here: https://github.com/calcom/cal.com/issues/12289#issuecomment-1837329444
This PR exposes a script for generating the openapi json and yaml files and putting them in the public directory.

One more note: The seed scripts in the prisma package generate a dredd-data.json file. This is used for values that were already seeded but needed to be exposed to the dredd testing script. eg: apiKeys and user Ids associated with them.

## Type of change


## How should this be tested?

To run the tests make sure the API server is running on `localhost:3002`. Then run the following command in the API directory:

```
yarn test:dredd
```

The generated dredd-output.html file can be easier to navigate than the CLI.

## Mandatory Tasks


## Other Notes about the API

The largest issue with the spec seems to be a lack of example responses and requests. These would likely be essential to anyone trying to create an integration.

------ AUTH

In many places, a 401 is thrown when a user does not have access to an item. A 403 or a 404 would be preferable. 403 would leak that the requested entity exists but user just does not have access. 401 does this as well, because we can still fetch other things with our valid apiKey. 404 would obfuscate the existence of the entities.

------ RESPONSE CASING

Most of the responses have a snake cased entity prop. Some have camel-cased. Making this consistent would be ideal.

------ Availability

        The response is not what the @swagger section says it will be.


------ Destination Calendars

      ALL:
        - No credential is seeded for any users and no API exists to make it as part of Dredd script

      POST:
        - credentialId is required but is not in list of params. Seems like credentialId does not actually need to be passed on the body since it gets looked up in the credential table


------ Teams

      POST:
        - Spec does not call out required or optional body fields properly
          - slug is not required
          - isPrivate, hideBookATeamMember, brandColor, darkBrandColor, timeZone, and weekStart are required by the Zod validator, but not listed in spec
